Python/immutability.c

```diff
@@ -156,281 +239,877 @@ static bool is_c_wrapper(PyObject* obj){
+    while(PyList_Size(state->pending) > 0){
+        pop(state->pending);
+    }
+    while(PyList_Size(state->dfs) > 0){
+        pop(state->dfs);
+    }
+
+    Py_DECREF(state->dfs);
+    Py_DECREF(state->pending);
 }
 
-static inline void
-gc_list_init(PyGC_Head *list)
+void set_direct_rc(PyObject* obj)
 {
-    // List header must not have flags.
-    // We can assign pointer by simple cast.
-    list->_gc_prev = (uintptr_t)list;
-    list->_gc_next = (uintptr_t)list;
+#ifndef GIL_DISABLED
+#if SIZEOF_VOID_P > 4
+    obj->ob_flags = (obj->ob_flags & ~_Py_IMMUTABLE_MASK) | _Py_IMMUTABLE_DIRECT;
+#else
+    obj->ob_refcnt = (obj->ob_refcnt & ~_Py_IMMUTABLE_FLAG) | _Py_IMMUTABLE_DIRECT;
+#endif
+#else
+    (void)obj;
+#endif
 }
 
-static inline int
-gc_list_is_empty(PyGC_Head *list)
+void set_indirect_rc(PyObject* obj)
 {
-    return (list->_gc_next == (uintptr_t)list);
+#ifndef GIL_DISABLED
+#if SIZEOF_VOID_P > 4
+    obj->ob_flags = (obj->ob_flags & ~_Py_IMMUTABLE_MASK) | _Py_IMMUTABLE_INDIRECT;
+#else
+    obj->ob_refcnt = (obj->ob_refcnt & ~_Py_IMMUTABLE_MASK) | _Py_IMMUTABLE_INDIRECT;
+#endif
+#else
+    (void)obj;
+#endif
 }
 
-/* Append `node` to `list`. */
-static inline void
-gc_list_append(PyGC_Head *node, PyGC_Head *list)
+bool has_direct_rc(PyObject* obj)
 {
-    assert((list->_gc_prev & ~_PyGC_PREV_MASK) == 0);
-    PyGC_Head *last = (PyGC_Head *)list->_gc_prev;
+#ifdef GIL_DISABLED
+    return false;
+#elif SIZEOF_VOID_P > 4
+    return (obj->ob_flags & _Py_IMMUTABLE_MASK) == _Py_IMMUTABLE_DIRECT;
+#else
+    return (obj->ob_refcnt & _Py_IMMUTABLE_FLAG) == _Py_IMMUTABLE_DIRECT;
+#endif
+}
 
-    // last <-> node
-    _PyGCHead_SET_PREV(node, last);
-    _PyGCHead_SET_NEXT(last, node);
 
-    // node <-> list
-    _PyGCHead_SET_NEXT(node, list);
-    list->_gc_prev = (uintptr_t)node;
+int is_representative(PyObject* obj, struct FreezeState *state)
+{
+#ifdef GIL_DISABLED
+    void* result = _Py_hashtable_get(state->rep, obj);
+    return ((uintptr_t)result & REPRESENTATIVE_FLAG) != 0;
+#else
+    return (_Py_AS_GC(obj)->_gc_prev & SCC_RANK_FLAG) != 0;
+#endif
 }
 
-/* Move `node` from the gc list it's currently in (which is not explicitly
- * named here) to the end of `list`.  This is semantically the same as
- * gc_list_remove(node) followed by gc_list_append(node, list).
- */
-static void
-gc_list_move(PyGC_Head *node, PyGC_Head *list)
+void set_scc_parent(PyObject* obj, PyObject* parent)
 {
-    /* Unlink from current list. */
-    PyGC_Head *from_prev = GC_PREV(node);
-    PyGC_Head *from_next = GC_NEXT(node);
-    _PyGCHead_SET_NEXT(from_prev, from_next);
-    _PyGCHead_SET_PREV(from_next, from_prev);
+    PyGC_Head* gc = _Py_AS_GC(obj);
+    // Use GC space for the parent pointer.
+    assert(((uintptr_t)parent & ~_PyGC_PREV_MASK) == 0);
+    uintptr_t finalized_bit = gc->_gc_prev & _PyGC_PREV_MASK_FINALIZED;
+    gc->_gc_prev = finalized_bit | _Py_CAST(uintptr_t, parent);
```

Comment:
Are you checking anywhere that the object is actually GC? Same for the functions below?

`_PyObject_IS_GC()` would be the function for this. All SCC objects should also be GC objects, but if this is not the case we should not assume that they are and overwrite unknown data here.

---

Edit: There is a check in `scc_init()` but this is just an assert and no proper handling of this